src/tests/end-to-end/common/mv3-manifest-instance.ts

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import * as fs from 'fs';
+import * as util from 'util';
+import { ManifestInstance } from 'tests/end-to-end/common/manifest-instance';
+
+export class ManifestV3Instance implements ManifestInstance {
+    private static readFile = util.promisify(fs.readFile);
+    private static writeFile = util.promisify(fs.writeFile);
+
+    public static async parse(manifestPath: string): Promise<chrome.runtime.ManifestV3> {
+        const buffer = await ManifestV3Instance.readFile(manifestPath);
+        const parsedManifest: chrome.runtime.ManifestV3 = JSON.parse(buffer.toString());
+        return parsedManifest;
+    }
+
+    public constructor(private readonly content: chrome.runtime.ManifestV3) {}
+
+    public addTemporaryPermission(permissionToAdd: string): ManifestV3Instance {
+        if (this.content.host_permissions == null) {
+            this.content.host_permissions = [];
+        }
+        this.content.host_permissions.push(permissionToAdd);
```

Comment:
should this be optional_host_permissions?

Reply:
My understanding here is that when we call `addTemporaryPermissions` in e2e tests we're essentially mocking the user giving us extra permissions. So in this case I think we would want it to be non-optional.